### PR TITLE
rewrite using operator-sdk

### PR DESCRIPTION
**Pull request by hensur**
Initially, I just wanted to add a new feature to the operator.
This change does much more.
While scanning the code base, I noticed, that editing a secret, for
example adding a new filed to the list of to-be-generated fields, can
lead to the regeneration of every field in the secret.

Also, the error returned by rand.Read wasn't checked.

Furthermore, the client-go used by this project was really old, instead
of just bumping the version, I decided it might be better to just
rewrite this using the operator-sdk.

I also added some integration tests, to make sure the operator behaves
as it's supposed to.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,36 +5,60 @@ os: linux
...
+        - diff -u <(echo -n) <(gofmt -d -s .)
+      name: "gofmt"
+
+    - stage: build
+      before_script:
+        - *before_build
+      script:
+        - make build DOCKER_IMAGE="${IMAGE_NAME}:latest" KUBE_SECRET_OPERATOR_VERSION=${TRAVIS_COMMIT::8}
+
+    - stage: deploy
```
Comment (martin-helmich) [inline .travis.yml:47]: Even with your change from 7ae5475, this job still seems to get triggered.

File: .travis.yml [hunk 2]
```diff
@@ -5,36 +5,60 @@ os: linux
...
+      install:
+        - curl -Lo kind https://github.com/kubernetes-sigs/kind/releases/download/${KIND_VERSION}/kind-linux-amd64 && chmod +x kind && sudo mv kind /usr/local/bin/
+        - kind get kubeconfig --internal --name kind-${CI_CONCURRENT_ID} | tee ${KUBECONFIG}
+      script:
+        - make test
+      name: "test"
+
+    - script:
+        - diff -u <(echo -n) <(gofmt -d -s .)
+      name: "gofmt"
```
Comment (martin-helmich) [inline .travis.yml:39]: We can just use CodeClimate for all the code styling stuff.

File: deploy/helm-chart/kubernetes-secret-generator/templates/deployment.yaml [hunk 3]
```diff
@@ -3,42 +3,55 @@ kind: Deployment
...
+            - name: POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: OPERATOR_NAME
+              value: "kubernetes-secret-generator"
+            - name: REGENERATE_INSECURE
+              value: {{ .Values.regenerateInsecure }}
+            - name: SECRET_LENGTH
+              value: {{ .Values.secretLength }}
```
Comment (martin-helmich) [inline deploy/helm-chart/kubernetes-secret-generator/templates/deployment.yaml:43]: I'm not sure how pedantic Kubernetes' validation is at this place, but `value` is defined as `string`. Just to be sure:

```suggestion
              value: {{ .Values.secretLength | quote }}
```

File: deploy/helm-chart/kubernetes-secret-generator/values.yaml [hunk 4]
```diff
@@ -43,3 +41,15 @@ nodeSelector: {}
...
+# are not cryptographically secure
+regenerateInsecure: "true"
+
+# Length of the generated secrets
+secretLength: 40
+
+# Namespace that are watched for secret generation
+# Accepts a comma-separated list of namespaces: ns1,ns2
+# If set to "", all namespaces will be watched
+watchNamespace: "default"
```
Comment (martin-helmich) [inline deploy/helm-chart/kubernetes-secret-generator/values.yaml:55]: The previous default behaviour was to watch all namespaces:

https://github.com/mittwald/kubernetes-secret-generator/blob/e0dbe6c8417d50c5ceaae3a9d93587591fdecb77/deploy/helm-chart/kubernetes-secret-generator/values.yaml#L6-L9

I'd prefer keeping this behaviour as to not introduce a massive breaking change to unsuspecting users.

File: pkg/controller/secret/secret_controller.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	SecretRegenerateAnnotation  = "secret-generator.v1.mittwald.de/regenerate"
+	SecretSecureAnnotation      = "secret-generator.v1.mittwald.de/secure"
+)
+
+func regenerateInsecure() bool {
+	return viper.GetBool("regenerate-insecure")
+}
+
+func secretLength() int {
+	return viper.GetInt("secret-length")
```
Comment (martin-helmich) [inline pkg/controller/secret/secret_controller.go:37]: Maybe introduce a default value (alternatively, abort with error), if this turns out to be `0`?

Comment (hensur) [reply pkg/controller/secret/secret_controller.go:37]: This is bound to the pFlags defined in `cmd/manager/main.go:64`. The pFlag has a default secret length of `40`. However, I guess we could still panic if this is manually set to 0.

File: deploy/helm-chart/kubernetes-secret-generator/values.yaml (hunk 4 above)
Comment (hensur) [reply deploy/helm-chart/kubernetes-secret-generator/values.yaml]: You're right, I changed the default in the helm chart.

**Pull request merged**